community/detectors/xwiki_cve_2024_21650/src/test/java/com/google/tsunami/plugins/detectors/rce/cve202421650/Cve202421650DetectorTest.java

```diff
@@ -0,0 +1,160 @@
+            .setServiceName("http")
+            .build();
+  }
+
+  @After
+  public void tearDown() throws IOException {
+    mockWebServer.shutdown();
+  }
+
+  @Test
+  public void detect_whenVulnerable_returnsVulnerability() throws IOException {
+    Cve202421650Detector mock = spy(detector);
+    when(mock.buildRandomString()).thenReturn(PSEUDO_RANDOM_STR);
+
+    mockWebServer.enqueue(
+        new MockResponse().setResponseCode(HttpStatus.OK.code()).setBody(CSRF_TEMPLATE));
+    mockWebServer.enqueue(
+        new MockResponse()
+            .setResponseCode(HttpStatus.OK.code())
+            .setBody("XWiki.test" + PSEUDO_RANDOM_STR + "]] (test" + PSEUDO_RANDOM_STR + ")"));
+    mockWebServer.enqueue(new MockResponse().setResponseCode(HttpStatus.ACCEPTED.code()));
+    mockCallbackServer.enqueue(PayloadTestHelper.generateMockSuccessfulCallbackResponse());
+
+    DetectionReportList detectionReports = mock.detect(targetInfo, ImmutableList.of(service));
+
+    assertThat(detectionReports.getDetectionReportsList())
+        .containsExactly(mock.buildDetectionReport(targetInfo, service));
+  }
+
+  @Test
+  public void detect_whenVulnerable_noCallbackServer_returnsVulnerability() {
+    Guice.createInjector(
+            new FakeUtcClockModule(fakeUtcClock),
+            new HttpClientModule.Builder().build(),
+            FakePayloadGeneratorModule.builder().build(),
+            new Cve202421650DetectorBootstrapModule())
+        .injectMembers(this);
+
+    Cve202421650Detector mock = spy(detector);
+    when(mock.buildRandomString()).thenReturn(PSEUDO_RANDOM_STR);
+
+    mockWebServer.enqueue(
+        new MockResponse().setResponseCode(HttpStatus.OK.code()).setBody(CSRF_TEMPLATE));
+    mockWebServer.enqueue(
+        new MockResponse()
+            .setResponseCode(HttpStatus.OK.code())
+            .setBody("XWiki.test" + PSEUDO_RANDOM_STR + "]] (test" + PSEUDO_RANDOM_STR + ")"));
+    mockWebServer.enqueue(new MockResponse().setResponseCode(HttpStatus.ACCEPTED.code()));
+
+    DetectionReportList detectionReports = mock.detect(targetInfo, ImmutableList.of(service));
+
+    assertThat(detectionReports.getDetectionReportsList())
+        .containsExactly(mock.buildDetectionReport(targetInfo, service));
+  }
+
+  @Test
+  public void detect_ifNotVulnerable_doesNotReportVuln() {
+    mockWebServer.enqueue(new MockResponse().setResponseCode(200).setBody(CSRF_TEMPLATE));
+    mockWebServer.enqueue(
+        new MockResponse()
+            .setResponseCode(HttpStatus.OK.code())
+            .setBody("<!DOCTYPE html><html><head></head><body>...</body></html>"));
+    mockWebServer.enqueue(new MockResponse().setResponseCode(HttpStatus.ACCEPTED.code()));
+    mockCallbackServer.enqueue(PayloadTestHelper.generateMockUnsuccessfulCallbackResponse());
+
+    DetectionReportList detectionReports = detector.detect(targetInfo, ImmutableList.of(service));
+    assertThat(detectionReports.getDetectionReportsList()).isEmpty();
+  }
```

Comment:
I think you should split this test in two different tests for the negative cases too (whether the callback server is available or not), like you did with the positive cases.

Reply:
changed